NitroxPatcher/Patches/Dynamic/AttackCyclops_UpdateAggression_Patch.cs

```diff
@@ -0,0 +1,119 @@
+                                                new CodeMatch(OpCodes.Ldsfld),
+                                                new CodeMatch(OpCodes.Callvirt, Reflect.Property((Player t) => t.currentSub).GetGetMethod())
+                                            ]).RemoveInstructions(25)
+                                            .InsertAndAdvance([
+                                                new CodeInstruction(OpCodes.Ldarg_0),
+                                                new CodeInstruction(OpCodes.Call, Reflect.Method(() => FindClosestCyclopsNoiseManagerIfAny(default)))
+                                            ]).InstructionEnumeration();
+    }
+
+    public static bool Prefix(AttackCyclops __instance)
+    {
+        return !__instance.TryGetNitroxId(out NitroxId creatureId) ||
+               Resolve<SimulationOwnership>().HasAnyLockType(creatureId);
+    }
+
+    public static void Postfix(AttackCyclops __instance)
+    {
+        if (__instance.currentTarget && __instance.currentTarget.TryGetNitroxId(out NitroxId targetId) &&
+            __instance.TryGetNitroxId(out NitroxId creatureId) && Resolve<SimulationOwnership>().HasAnyLockType(creatureId))
+        {
+            float aggressiveToNoise = __instance.aggressiveToNoise.Value;
+
+            Resolve<IPacketSender>().Send(new AttackCyclopsTargetChanged(creatureId, targetId, aggressiveToNoise));
+        }
+    }
+
+    /// <summary>
+    /// This is executed every 0.5s for each spawned leviathan so we can focus on optimization
+    /// </summary>
+    public static CyclopsNoiseManager FindClosestCyclopsNoiseManagerIfAny(AttackCyclops attackCyclops)
+    {
+        // Limit for optimization
+        if (NitroxCyclops.ScaledNoiseByCyclops.Count > 100)
+        {
+            // Cyclops are marked with EcoTargetType.Whale
+            IEcoTarget ecoTarget = EcoRegionManager.main.FindNearestTarget(EcoTargetType.Whale, attackCyclops.transform.position, IsTargetAValidInhabitedCyclops, 3);
+            if (ecoTarget == null)
+            {
+                return null;
+            }
+            
+            return ecoTarget.GetGameObject().GetComponent<CyclopsNoiseManager>();
+        }
+        
+        float minDistance = float.MaxValue;
+        CyclopsNoiseManager closest = null;
+        foreach (KeyValuePair<NitroxCyclops, float> cyclopsEntry in NitroxCyclops.ScaledNoiseByCyclops)
+        {
+            if (cyclopsEntry.Key.Pawns.Count == 0)
+            {
+                continue;
+            }
+
+            // Calculations from the "if (closestDecoy != null || cyclopsNoiseManager != null)" part
+            float distance = Vector3.Distance(cyclopsEntry.Key.transform.position, attackCyclops.transform.position);
+            
+            if (distance < cyclopsEntry.Value && distance < minDistance)
+            {
+                minDistance = distance;
+                closest = cyclopsEntry.Key.GetComponent<CyclopsNoiseManager>();
```

Comment:
I would only save  `cyclopsEntry.Key` and do the GetComponent part later so we don't do unnecessary heavy unity calls if `NitroxCyclops.ScaledNoiseByCyclops` has an unfortunate sorting.